ssl/ssl_transfer_asn1.cc

```diff
@@ -335,18 +544,29 @@ static int SSL3_STATE_parse_session(CBS *cbs, UniquePtr<SSL_SESSION> *out,
 // |ssl| is used because |tls1_configure_aead| is used to recover
 // |aead_read_ctx| and |aead_write_ctx|.
 static int SSL3_STATE_from_bytes(SSL *ssl, CBS *cbs, const SSL_CTX *ctx) {
+  // We expect the caller to have configured |ssl| with the protocol
+  // version prior to calling us.
+  uint16_t protocol_version;
+  if (!ssl_protocol_version_from_wire(&protocol_version, ssl->version)) {
+    OPENSSL_PUT_ERROR(SSL, SSL_R_SERIALIZATION_INVALID_SSL3_STATE);
+    return 0;
+  }
+  bool is_tls13 = protocol_version >= TLS1_3_VERSION;
+
   SSL3_STATE *out = ssl->s3;
   CBS s3, read_seq, write_seq, server_random, client_random, send_alert,
       pending_app_data, read_buffer;
   CBS previous_client_finished, previous_server_finished;
   int session_reused, channel_id_valid, send_connection_binding, not_resumable;
-  uint64_t version, early_data_reason, previous_client_finished_len,
+  uint64_t serde_version, early_data_reason, previous_client_finished_len,
       previous_server_finished_len;
   uint64_t empty_record_count, warning_alert_count, total_renegotiations;
   int64_t rwstate;
   int pending_app_data_present, read_buffer_present;
   if (!CBS_get_asn1(cbs, &s3, CBS_ASN1_SEQUENCE) ||
-      !CBS_get_asn1_uint64(&s3, &version) || version != kS3Version ||
+      !CBS_get_asn1_uint64(&s3, &serde_version) ||
+      serde_version > SSL3_STATE_SERDE_VERSION_TWO ||
+      (is_tls13 && serde_version < SSL3_STATE_SERDE_VERSION_TWO) ||
```

Comment:
should factor out validation checks into its own function at some point I guess.